inc/Result.h

```diff
@@ -0,0 +1,24 @@
+#ifndef RESULT_H
+#define RESULT_H
+
+#include <string>
+
+template <typename T>
+class Result
```

Comment:
I think result types are a good idea in situations if multiple different types of failures could happen. Like you mentioned in your update message for Friday, though, it's more complicated than just returning a nullptr on failure, but it does give some more information about the nature of the failure. I think it is a better design choice, but there's definitely a tradeoff because it complicates things.